src/System.Net.Quic/src/System/Net/Quic/Implementations/QuicConnectionProvider.cs

```diff
@@ -0,0 +1,30 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Quic.Implementations
+{
+    internal abstract class QuicConnectionProvider : IDisposable
```

Comment:
Don't these need to be public if we're going to use them to implement prototypes outside of the repo?

Reply:
I'm also still going to advocate for having QuicConnection be virtual/abstract over this provider pattern. What concrete benefits are there to having the provider pattern over letting us reimplement these types ourselves? Easier to take out later?

Reply:
Yes, easier to take out later is a big part of it. We've definitely had situations in the past where "temporary" APIs have turned into de facto permanent APIs because they were used and then it became too difficult to remove them. So the more we can do to avoid this, the better.

In general we want to focus on producing the right public API here. Doing a bunch of temporary things in the public API is counterproductive to that.

Reply:
> Don't these need to be public if we're going to use them to implement prototypes outside of the repo?

Why do we want to do this? My assumption is that prototypes should be in the repo, either in a separate proto branch or in master, depending on the work.

Reply:
If you want to test Kestrel with this abstraction, we wouldn't be able to. Why not allow this extensibility to be used for InMemory?

Reply:
I don't want to test Kestrel with this abstraction. This abstraction is going away before we ship. I assume we want Kestrel tests to keep working once we remove this abstraction.

Reply:
To be clear, I think that as a point in time thing we may want to be able to run some Kestrel tests using different providers, including the mock provider. But that's a temporary thing, right?

Reply:
Is this how we test a prototype against Kestrel?
- Merge it into corefx
- Wait for a dependency update
- New up a provider and pass it into QuicListener/Connection?

And if we find a bug in the prototype we have to go all the way through that flow again?

Having to merge into corefx in order to test prototypes with Kestrel is not a productive flow. We should be able to implement the prototypes in their own repos and test them with corefx or kestrel without ever merging them in.

Note some of these prototypes may involve 3rd party licenses that we do not want to merge into any of our repos.

Reply:
> Is this how we test a prototype against Kestrel?

Today we have two separate repos for coreclr and corefx, with coreclr consumed into corefx via dependency updates.  But I can have both repos locally and use my local corefx against my local coreclr, without needing to wait for any dependency update in the middle.  We should be able to do exactly the same thing between dotnet/runtime and dotnet/aspnetcore; in fact, it should be even easier, because part of the goal of dotnet/runtime is to produce a fully-functional shared framework.

> Having to merge into corefx in order to test prototypes with Kestrel is not a productive flow. 

QUIC is not simple.  How many full-fledged prototypes are you going to be creating?  And why wouldn't we want to test out such prototypes on the client in addition to on the server?  Why would we want such prototypes living in completely separate places, likely duplicating code and effort?  Why would we want to move those implementations between repos when we decide on the one we actually want to ship?  Etc.

Reply:
I don't understand why we don't have the will power to design abstractions. Even if we choose not to ship them, we're a year out from shipping 5.0 RTM. Can't we have some decent public abstractions in place for testing? Sure it'll take some time but the benefits are huge (and we'll get some design mileage for free!).

Why is this a problem?

Reply:
> I don't understand why we don't have the will power to design abstractions.

No one is saying we can't have abstractions. What we are saying is that we should be conservative and deliberate about designing and adding abstractions, as we always are for corefx.

Designing good abstractions is hard, and shipping bad abstractions has significant costs for a long time. We know this from experience.

Let's continue to have the conversation about what an abstraction here should look like. As you say, we have plenty of time for 5.0 to figure this out. But in the meantime, let's make progress on QUIC implementation and let's not block on figuring out abstractions.

Reply:
@Tratcher, I expect the "flow" can be much simpler than that.

Implementing a prototype, or testing a bug fix or a perf improvement or whatever, can happen in a feature branch or a private branch or whatever. To test with it, you just need to grab the private corefx bits (either the full framework, or just System.Net.Quic.dll) and use it.

Doing this sort of prototyping in corefx allows us to easily test with either HttpClient or Kestrel or some other QUIC-using component. It allows us to easily run functional, stress, and perf tests as we build these out. It allows us to easily merge these changes into master if we decide we want to take them.

Reply:
We all understand the cost of shipping bad APIs in general (abstraction or not). In fact if HttpClient used an abstraction we wouldn’t have any of these problems. Quic itself wouldn’t need a public abstraction (if we’re against that for some reason because it a new surface area)

Reply:
> In fact if HttpClient used an abstraction we wouldn’t have any of these problems. Quic itself wouldn’t need a public abstraction

We're adding one:
https://github.com/dotnet/corefx/issues/35404
So it sounds like we're in agreement no public QUIC abstraction is required.

Reply:
Sounds great to me!